core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/NormalizeOrInExpressionRewriter.java

```diff
@@ -0,0 +1,122 @@
+import com.google.common.collect.ImmutableSet;
+import io.trino.sql.tree.ComparisonExpression;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.ExpressionRewriter;
+import io.trino.sql.tree.ExpressionTreeRewriter;
+import io.trino.sql.tree.InListExpression;
+import io.trino.sql.tree.InPredicate;
+import io.trino.sql.tree.LogicalExpression;
+
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.ExpressionUtils.and;
+import static io.trino.sql.ExpressionUtils.or;
+import static io.trino.sql.tree.ComparisonExpression.Operator.EQUAL;
+import static io.trino.sql.tree.LogicalExpression.Operator.AND;
+
+public final class NormalizeOrInExpressionRewriter
+{
+    public static Expression normalizeOrInExpression(Expression expression)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression);
+    }
+
+    private NormalizeOrInExpressionRewriter() {}
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        @Override
+        public Expression rewriteLogicalExpression(LogicalExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            List<Expression> terms = node.getTerms().stream()
+                    .map(expression -> treeRewriter.rewrite(expression, context))
+                    .collect(toImmutableList());
+
+            if (node.getOperator() == AND) {
+                return and(terms);
+            }
+
+            ImmutableMultimap.Builder<Expression, Expression> expressionBuilder = ImmutableMultimap.builder();
+            for (Expression expression : terms) {
+                if (expression instanceof ComparisonExpression comparisonExpression && comparisonExpression.getOperator() == EQUAL) {
+                    expressionBuilder.put(comparisonExpression.getLeft(), comparisonExpression);
+                }
+                else if (expression instanceof InPredicate inPredicate && inPredicate.getValueList() instanceof InListExpression) {
+                    expressionBuilder.put(inPredicate.getValue(), inPredicate);
+                }
+            }
+
+            ImmutableList.Builder<InPredicate> inPredicateBuilder = ImmutableList.builder();
+            ImmutableSet.Builder<Expression> expressionToSkipBuilder = ImmutableSet.builder();
+            expressionBuilder.build().asMap().forEach((expression, values) -> {
+                if (values.size() > 1) {
+                    inPredicateBuilder.add(new InPredicate(expression, mergeToInListExpression(values)));
+                    expressionToSkipBuilder.add(expression);
+                }
+            });
+
+            Set<Expression> expressionToSkip = expressionToSkipBuilder.build();
+            ImmutableList.Builder<Expression> othersExpressionBuilder = ImmutableList.builder();
```

Comment:
I think `othersExpressionBuilder` should be build next to `expressionBuilder` as you are using similar conditions (repeat the logic) but negated.